packages/core/src/transport/sendWithRetryStrategy.ts

```diff
@@ -0,0 +1,153 @@
+import { addTelemetryDebug } from '../domain/telemetry'
+import { monitor } from '../tools/monitor'
+import { ONE_KILO_BYTE, ONE_MEGA_BYTE, ONE_SECOND } from '../tools/utils'
+import type { Payload, HttpResponse } from './httpRequest'
+
+export const MAX_ONGOING_BYTES_COUNT = 80 * ONE_KILO_BYTE
+export const MAX_ONGOING_REQUESTS = 32
+export const MAX_QUEUE_BYTES_COUNT = 3 * ONE_MEGA_BYTE
+export const MAX_BACKOFF_TIME = 256 * ONE_SECOND
+
+export interface RetryState {
+  isIntakeAvailable: boolean
+  currentBackoffTime: number
+  bandwidthMonitor: ReturnType<typeof newBandwidthMonitor>
+  queuedPayloads: ReturnType<typeof newPayloadQueue>
+}
+
+type SendStrategy = (payload: Payload, onResponse: (r: HttpResponse) => void) => void
+
+export function sendWithRetryStrategy(payload: Payload, state: RetryState, sendStrategy: SendStrategy) {
+  if (state.isIntakeAvailable && state.bandwidthMonitor.canHandle(payload)) {
+    send(payload, state, sendStrategy, {
+      onSuccess: () => sendNextPayload(state, sendStrategy),
+      onFailure: () => {
+        state.queuedPayloads.enqueue(payload)
+        scheduleRetry(state, sendStrategy)
+      },
+    })
+    sendNextPayload(state, sendStrategy)
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** Try to send the queued payloads (if any) before sending the new payload.

Currently, if a payload bigger than `MAX_ONGOING_BYTES_COUNT` is added to the queue, it wont ever be sent (assuming the intake always succeeds), because `bandwidthMonitor.ongoingRequestCount` will always be greater than `1` (since we just started sending a request) and the payload size always bigger than `MAX_ONGOING_BYTES_COUNT`, so `canHandle` will always return false.

Reply:
> it wont ever be sent (assuming the intake always succeeds), because bandwidthMonitor.ongoingRequestCount will always be greater than 1 (since we just started sending a request) and the payload size always bigger than MAX_ONGOING_BYTES_COUNT, so canHandle will always return false.


not exactly, we will wait for no more ongoing requests to send it, cf: https://github.com/DataDog/browser-sdk/blob/82c719f00de87b1c89e5ffeed18c13fbd2ca6eac/packages/core/src/transport/sendWithRetryStrategy.ts#L137-L143

Reply:
Hum I missed the `sendNextPayload` call in `onSuccess`. But I still think we should first try to send payloads from the queue before sending the new one, so the "first in first out" rule is respected. It could be useful in the Session Replay case where segments follow each other: if we miss an intermediary segment, the replay won't work, but if we miss the last one, it will only be truncated.

Reply:
Currently, we are maximizing our bandwidth consumption:
- if a new request can be sent, it is sent
- when a request complete, we try to send all possible queued requests than can fit the available bandwidth.
So it indeed may not respect the order when the bandwidth is under constraint.

Enforcing the order indeed make sense for session replay, we missed that during previous reflexion.
I'll have a look.

Reply:
Maybe it could be configured per intake? We should make sure the queue limit is big enough to avoid reaching it in "normal usage". This limit is only there to prevent memleaks (ex: when a page is continuously sending data to a failing intake for a long time), right?

Reply:
> We should make sure the queue limit is big enough to avoid reaching it in "normal usage". This limit is only there to prevent memleaks (ex: when a page is continuously sending data to a failing intake for a long time), right?

yes, the reasoning for removing oldest elements was "if they are very old, they will be ignored by the backend anyway" 

> Maybe it could be configured per intake?

given the original reasoning I don't think it worth the trouble

Reply:
How about adding some telemetry to see if it happens often, and raise the limit if that's the case?

Reply:
is it still related to choosing between discarding oldest or newest queued payloads?
or is it another point?

Reply:
It is to evaluate how much it matters. If the limit is reached very occasionnaly, let's keep discarding oldest payloads. Else, we should consider a different approach.

Reply:
I already made the change, it seems more consistent if we want to favor request order.
I'll add some monitoring, let me know if you think we should get back to discarding oldest payloads.